horovod/tensorflow/__init__.py

```diff
@@ -330,10 +330,9 @@ def after_create_session(self, session, coord):
             session.run(self.bcast_op)
 
 
-@_cache
 def _make_allreduce_grads_fn(name, device_dense, device_sparse,
                              compression, sparse_as_dense, op, gradient_predivide_factor,
-                             num_groups):
+                             num_groups, group_lists=None):
```

Comment:
Hey @zhuzilin, instead of removing the `@_cache`, you should be able to make the `group_lists` hashable using a simple trick.  First we can define these helper functions in `horovod/tensorflow/util.py`:

```
def vars_to_refs(vars):
    return tuple(v.ref() for v in vars)

def refs_to_vars(refs):
    return [r.deref() for r in refs]
```

Then we can use these functions to wrap the `_make_allreduce_grads_fn`:

```
@_cache
def _make_cached_allreduce_grads_fn(..., groups):
    groups = refs_to_vars(group_refs) if isinstance(groups, tuple) else groups
    ...

def _make_allreduce_grads_fn(...):
    groups = vars_to_refs(groups) if isinstance(groups, list) else None
    return _make_allreduce_grads_fn(..., groups)
```

Does that make sense?